src/app/authentication.ts

```diff
@@ -1,139 +1,268 @@
+
+        resultRequestCounter += 1;
+
+        if (req.body.response?.authenticatorData) {
+            try {
+                const authenticatorData = parseAuthenticatorData(Buffer.from(
+                    req.body.response?.authenticatorData, "base64"));
+                const authenticatorFlagsUP = authenticatorData?.flags?.up;
+                const authenticatorFlagsUV = authenticatorData?.flags?.uv;
+
+                /**
+                 * According to webauthn specification, the server should check for the existance of UP flag.
+                 * Quaoted from spec: "Verify that the User Present bit of the flags in authData is set".
+                 * Hence this scenario is explicitly handled from the adapter.
+                 */
+                if (!authenticatorFlagsUP) {
+                    res.send({
+                        errorMessage: "",
+                        status: "ok"                        
+                    });
+
+                    return;
+                }
+
+                /**
+                 * We are handling this test case explicitly since the server doesn't allow to config user verification
+                 * option. According to our design, this option cannnot be configured and will always be set to
+                 * "preferred".
+                 */
+                if (authenticatorFlagsUP && !authenticatorFlagsUV && userVerification == "required") {
+                    res.send({
+                        errorMessage: "",
+                        status: "failed"
+                    });
+
+                    return;
+                }
+            } catch (e) {
+                console.log("Error in decoding authenticatorData object!");
+            }
         }
-      )
-      .then(async (response) => {
-        if (response.request.path == "/pickup-dispatch/index.jsp") {
-          res.send({
-            status: "failed",
-            errorMessage: "",
-          });
-        } else {
-          res.send({
-            status: "ok",
-            errorMessage: "",
-          });
+
+        /**
+         * These parameters are not supported by the yubico data structure. Therefore need to remove before 
+         * sending to the backend implementation. Otherwise will throw data conversion exception.
+         */
+        req.body["clientExtensionResults"] = {};
+        delete req.body["getClientExtensionResults"];
+
+        if (req.body["authenticatorAttachment"]) {
+            delete req.body["authenticatorAttachment"];
         }
-      })
-      .catch((error) => {
+
+        const tr = JSON.stringify({ credential: req.body, requestId: requestId });
+        const referer = `https://${config.authRequestRefererHost}`
+            + (config.tenantName && config.tenantName !== "" ? `/t/${config.tenantName}/` : "/") 
+            + "authenticationendpoint/fido2-auth.jsp?authenticators=FIDOAuthenticator%3ALOCAL"
+            + `&type=fido&sessionDataKey=${sessionDataKey}&data=${tr}`;
```

Comment:
Shall we use one of templates literals or double quotes only?

Reply:
Fixed